### PR TITLE
feat: [Typography] Add <Numeral /> component for number formatting

### LINKED ISSUE #1136: [Typography] Add `<Numeral />` component for number formatting
### Which Component Need Enhancement 期望新增功能的组件
- Typography

### What does the proposed API look like 期望支持的API
Typography.Numeral

### Feature Description 功能描述
Custom number formats can control the display of numbers, dates, times, fractions, percentages, and other numeric values. 

It might look like this when used:
``` javascript
import React from 'react';
import { Typography } from '@douyinfe/semi-ui';

function Demo() {
    const { Numeral } = Typography;
    return (
        <div>  
            <Numeral rule='currency'  precision={2}>
                   1000.2 ¥
            </Numeral>
            <br />
            <Numeral rule='bytes-binary'>
                   1024
            </Numeral>
            <br />
            <Numeral rule='exponential'>
                   1123456789
            </Numeral>
        </div>
    );
}
```

``` text
1,000.20 ¥

1KiB

1e+9
```

It also inherits the attribute of <Text />.
In addition it is possible to customise the formatting rules method:
``` javascript
<Numeral 
    type='warning' 
    copyable={true} 
    parser={(oldValue:string):string => { ...; return newValue }}
>
      1+1
</Numeral>
```

### Additional information 补充说明
I applied semi to the backend system and found it unfortunate that `<InputNumber />` was supported but not the formatted presentation of number.

However, I'm not sure if my judgement is accurate or not, so I'll need to trouble you to judge and if appropriate I'll still try to add this content for reference.

### PR DESCRIPTION
**Pull request by uiuing**
## What kind of change does this PR introduce? (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Test Case
- [ ] TypeScript definition update
- [ ] Document improve
- [ ] CI/CD improve
- [ ] Branch sync
- [ ] Other, please describe:

  ## PR description

Fixes #1136

## Changelog
🇨🇳 Chinese
- Feat: 新增<Numeral />组件，基于Text组件，添加了属性: rule, precision, truncate, parser, 以提供需要单独处理文本中数值的能力

---

🇺🇸 English
- Feat: Add <Numeral /> component, Based on Text component, added properties: rule, precision, truncate, parser, to provide the Numeral to handle values in text separately

## Checklist
- [x] Test or no need
- [x] Document or no need
- [ ] Changelog or no need

## Other
- [ ] Skip Changelog

### (1) Introduction to test
Tested the use of the added attributes

File: [typography.test.js](https://github.com/DouyinFE/semi-design/compare/main...uiuing:semi-design:main#diff-83c5594804fd5dcf0bf2c9d6d58e9c2c250fafb29a057d001b20e855f17273e9)

### (2) Introduction to document
File: [index.md](https://github.com/DouyinFE/semi-design/compare/main...uiuing:semi-design:main#diff-4639cd3b5d23dd5bb6ead05474bf6f1e171d13f1ab975305928d9b618798410f) 、[index-en-us.md](https://github.com/DouyinFE/semi-design/compare/main...uiuing:semi-design:main#diff-455f827689aa8334cd41535f83ad89c52e833c4528cb22392733f35f3e0294d3)

image

image

### (3) Some considerations
Processing logic: 

Digital format reference from:
- [Byte - wikipedia](https://en.wikipedia.org/wiki/Byte)
- [Available number formats in Excel](https://support.microsoft.com/en-us/office/available-number-formats-in-excel-0afe8f52-97db-41f1-b972-4b46e9f1e8d2)

### (4) Possible problem areas
- There may be room for more optimisation of these codes
- Regarding the `rule` attributes supported by: `text`, `numbers`, `bytes-decimal`, `bytes-binary`, `percentages`, `currency`, `exponential` , I'm not sure if all these types are needed or missing something

Include these or those that may not have been mentioned, Request for inspection and correction

### CODE REVIEW

**Comment by pointhalo** [comment]
This feature is planned to be released with the next minor version🚀,  and will be reviewed before release

**Comment by uiuing** [comment]
> This feature is planned to be released with the next minor version🚀, and will be reviewed before release

Yay, If it needs to be changed I will do so immediately 🫡

File: content/basic/typography/index.md [hunk 1]
```diff
@@ -137,6 +137,48 @@ function Demo() {
...
+基于Text组件，添加了属性: `rule`, `precision`, `truncate`, `parser`, 以提供需要单独处理文本中数值的能力
+
+```jsx live=true
+import React from 'react';
+import { Typography } from '@douyinfe/semi-ui';
+
+function Demo() {
+    const { Numeral } = Typography;
+    return (
+        <div>
```
Comment (pointhalo) [inline content/basic/typography/index.md:151]: Any example for props.rule='numbers' and 'text' ?  I can't get their behavior from this document right now

File: packages/semi-foundation/typography/formatNumeral.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                return `${this.truncatePrecision(value * 100)}%`;
+            }
+            return `${this.truncatePrecision(value)}%`;
+        },
+        currency: (value: number) => {
+            const cArr = this.truncatePrecision(value).split('.');
+            const cInt = cArr[0].replace(/\d{1,3}(?=(\d{3})+$)/g, '$&,');
+            const cFloat = cArr[1] ? `.${cArr[1]}` : '';
+            return `${cInt}${cFloat}`;
+        },
```
Comment (pointhalo) [inline packages/semi-foundation/typography/formatNumeral.ts:55]: Currency display rules, different regions have different customs, separators and the rules for inserting separators may be different, not international. It is not possible to simply adapt directly with a thousandth comma.
eg:
     - 🇷🇺 371,10 руб.
     - 🇮🇳‎ ₹1,04,990
     - 🇮🇩 Rp124.000

If we decide to add rule=currency, our users should be expected to adapt to the delimiter habits of the region where the language set in LocaleProvider is located.
This feature can be a bit fiddly, and it takes some time to research the standards. At present, I have read the oracle documentation that mentions a part
https://docs.oracle.com/cd/E19253-01/819-0402/overview-52/index.html

I feel that this process will be longer, maybe we can open a separate PR to follow up. 
Can't add it for now, what do you think?

Comment (uiuing) [reply packages/semi-foundation/typography/formatNumeral.ts:55]: Oh! This was not fully thought out at first, and it's a real problem, so `rule=currency` will be discussed and added later

For the handling of USD how about I turn it into a case in the documentation?  😉

File: content/basic/typography/index.md (hunk 1 above)
Comment (uiuing) [reply content/basic/typography/index.md:151]: Roger, I will add it soon

Comment (pointhalo) [reply content/basic/typography/index.md:164]: image

i push a commit, added more detailed api usage,  you can continue to add here

**Pull request merged**